### PR TITLE
Add support for single reductions in ComplexDeinterleavingPass

### PR DESCRIPTION
**Pull request by NickGuy-Arm**
The Complex Deinterleaving pass assumes that all values emitted will result in complex numbers, this patch aims to remove that assumption and adds support for emitting just the real or imaginary components, not both.

### CODE REVIEW

File: llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp [hunk 1]
```diff
@@ -1511,6 +1690,7 @@ bool ComplexDeinterleavingGraph::collectPotentialReductions(BasicBlock *B) {
 }
 
 void ComplexDeinterleavingGraph::identifyReductionNodes() {
+  dbgs() << "identifyReductionNodes\n";
```
Comment (SamTebbs33) [inline llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp:1693]: I assume this has been left over from debugging.

File: llvm/test/CodeGen/AArch64/complex-deinterleaving-cdot.ll [hunk 2]
```diff
@@ -0,0 +1,170 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
```
Comment (SamTebbs33) [inline llvm/test/CodeGen/AArch64/complex-deinterleaving-cdot.ll:1]: Could you add a more focused test without the loop?

File: llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp [hunk 3]
```diff
@@ -891,18 +902,178 @@ ComplexDeinterleavingGraph::identifySymmetricOperation(Instruction *Real,
...
+
+  return nullptr;
+}
+
 ComplexDeinterleavingGraph::NodePtr
 ComplexDeinterleavingGraph::identifyNode(Value *R, Value *I) {
-  LLVM_DEBUG(dbgs() << "identifyNode on " << *R << " / " << *I << "\n");
-  assert(R->getType() == I->getType() &&
-         "Real and imaginary parts should not have different types");
+  bool _;
```
[the remark this answers is not in the document]
Comment (NickGuy-Arm) [reply llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp:1057]: For one of the sanity-checks elsewhere (https://github.com/llvm/llvm-project/pull/112875/files/b2410688146531936db5f58ed2f0ebf78bf8387a#diff-ebbbd6cbc055d2185b50e106f58ee13188b47cd5fa49f21fb66a9ea82d54b086R993), we check that we're pulling the already-matched node from the cache after unrolling any casts.
The bool here is to act as an optional parameter to `identifyNode`, working around the lvalue/rvalue mismatch of having it inlined.

It's either this approach, or we update every call to `identifyNode` to have an extra bool variable assigned nearby. (Or we remove the sanity check altogether)

File: llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp (hunk 1 above)
Comment (NickGuy-Arm) [reply llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp:1693]: Yep, I got all the ones from actual debugging, but not those that I had to add to debug a merge conflict. Removed.

File: llvm/test/CodeGen/AArch64/complex-deinterleaving-cdot.ll (hunk 2 above)
Comment (NickGuy-Arm) [reply llvm/test/CodeGen/AArch64/complex-deinterleaving-cdot.ll:1]: I've removed as much of the loop as I can, however we need the IR to at least resemble a loop in order for the reduction to be valid and recognised.

File: llvm/test/CodeGen/AArch64/complex-deinterleaving-cdot.ll [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+; RUN: opt -S --passes=complex-deinterleaving %s --mattr=+sve2 -o - | FileCheck %s
```
Comment (SamTebbs33) [inline llvm/test/CodeGen/AArch64/complex-deinterleaving-cdot.ll:2]: We need some more tests, such as:
* No sve, no sve2
* More rotations
* I don't really understand the caching system, but some tests that test that functionality are needed
* More types

And negative tests for other code paths.

**Comment by NickGuy-Arm** [comment]
Ping

File: llvm/test/CodeGen/AArch64/complex-deinterleaving-cdot.ll [hunk 5]
```diff
@@ -1,41 +1,100 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
-; RUN: opt -S --passes=complex-deinterleaving %s --mattr=+sve2 -o - | FileCheck %s
+; RUN: opt -S --passes=complex-deinterleaving %s --mattr=+sve2 -o - | FileCheck %s --check-prefix=CHECK-SVE2
```
[the remark this answers is not in the document]
Comment (NickGuy-Arm) [reply llvm/test/CodeGen/AArch64/complex-deinterleaving-cdot.ll:2]: I've added a negative test, but the caching is an implementation detail and is not exposed in any testable way, it's also not new in this PR

File: llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp [hunk 6]
```diff
@@ -891,17 +901,172 @@ ComplexDeinterleavingGraph::identifySymmetricOperation(Instruction *Real,
...
+  CN->addOperand(identifyNode(BReal, BImag));
+  CN->addOperand(identifyNode(Phi, RealUser));
+
+  return submitCompositeNode(CN);
+}
+
+ComplexDeinterleavingGraph::NodePtr
+ComplexDeinterleavingGraph::identifyPartialReduction(Value *R, Value *I) {
+  if (!I->hasOneUser())
+    return nullptr;
```
Comment (SamTebbs33) [inline llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp:1014]: Why can there only be one user? We should have a negative test for it so someone removing the check in the future doesn't think all is well!

File: llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp [hunk 7]
```diff
@@ -891,17 +901,172 @@ ComplexDeinterleavingGraph::identifySymmetricOperation(Instruction *Real,
...
+
+  if (NodePtr CN = identifyDotProduct(IInst))
+    return CN;
 
+  return nullptr;
+}
+
+ComplexDeinterleavingGraph::NodePtr
+ComplexDeinterleavingGraph::identifyNode(Value *R, Value *I, bool *FromCache) {
```
Comment (SamTebbs33) [inline llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp:1054]: Can `FromCache` be a reference instead?

File: llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp [hunk 8]
```diff
@@ -891,17 +901,172 @@ ComplexDeinterleavingGraph::identifySymmetricOperation(Instruction *Real,
...
+  if (!RealTy)
+    return nullptr;
+  VectorType *ImagTy = dyn_cast<VectorType>(I->getType());
+  if (!ImagTy)
+    return nullptr;
+
+  if (RealTy->isScalableTy() != ImagTy->isScalableTy())
+    return nullptr;
+  if (RealTy->getElementType() != ImagTy->getElementType())
+    return nullptr;
```
Comment (SamTebbs33) [inline llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp:1026]: We should have some tests for these.

File: llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp (hunk 6 above)
Comment (NickGuy-Arm) [reply llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp]: The `hasOneUser` check was to simplify things later, but it's not strictly necessary so I've removed it and fixed the succeeding code to not assume one user.

File: llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp (hunk 8 above)
Comment (NickGuy-Arm) [reply llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp]: Added some tests

File: llvm/test/CodeGen/AArch64/complex-deinterleaving-cdot.ll [hunk 9]
```diff
@@ -913,8 +913,224 @@ middle.block:                                     ; preds = %vector.body
...
+  %imag.mul.neg = sub <vscale x 16 x i32> zeroinitializer, %imag.mul
+  %partial.reduce.sub = call <vscale x 8 x i16> @llvm.experimental.vector.partial.reduce.add.nxv8i16.nxv16i32(<vscale x 8 x i16> %real.mul.reduced, <vscale x 16 x i32> %imag.mul.neg)
+  br i1 true, label %middle.block, label %vector.body
+
+middle.block:                                     ; preds = %vector.body
+  %0 = call i16 @llvm.vector.reduce.add.nxv8i16(<vscale x 8 x i16> %partial.reduce.sub)
+  ret i16 %0
+}
+
+define i32 @cdotp_i8_rot0_fixed_length(<32 x i8> %a, <32 x i8> %b) {
```
Comment (SamTebbs33) [inline llvm/test/CodeGen/AArch64/complex-deinterleaving-cdot.ll:1022]: This name makes it seem like a positive test so I'd suggest it be changed to `not_cdotp_fixed_length`.

**Review by SamTebbs33 — APPROVED** [review]
LGTM with one request.

**Comment by fhahn** [comment]
It looks like this may be causing some buildbot failures: https://lab.llvm.org/buildbot/#/builders/41/builds/4171


Could you please take a look and revert if it isn't a trivial fix

```
FAILED: lib/Demangle/CMakeFiles/LLVMDemangle.dir/MicrosoftDemangle.cpp.o 
/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage2/lib/Demangle -I/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/llvm/llvm/lib/Demangle -I/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage2/include -I/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/llvm/llvm/include -mcpu=neoverse-512tvb -mllvm -scalable-vectorization=preferred -mllvm -treat-scalable-fixed-error-as-warning=false -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/Demangle/CMakeFiles/LLVMDemangle.dir/MicrosoftDemangle.cpp.o -MF lib/Demangle/CMakeFiles/LLVMDemangle.dir/MicrosoftDemangle.cpp.o.d -o lib/Demangle/CMakeFiles/LLVMDemangle.dir/MicrosoftDemangle.cpp.o -c /home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/llvm/llvm/lib/Demangle/MicrosoftDemangle.cpp
clang++: ../llvm/llvm/lib/IR/Value.cpp:508: void llvm::Value::doRAUW(Value *, ReplaceMetadataUses): Assertion `New->getType() == getType() && "replaceAllUses of value with new value of different type!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage2/lib/Demangle -I/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/llvm/llvm/lib/Demangle -I/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage2/include -I/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/llvm/llvm/include -mcpu=neoverse-512tvb -mllvm -scalable-vectorization=preferred -mllvm -treat-scalable-fixed-error-as-warning=false -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17 -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/Demangle/CMakeFiles/LLVMDemangle.dir/MicrosoftDemangle.cpp.o -MF lib/Demangle/CMakeFiles/LLVMDemangle.dir/MicrosoftDemangle.cpp.o.d -o lib/Demangle/CMakeFiles/LLVMDemangle.dir/MicrosoftDemangle.cpp.o -c /home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/llvm/llvm/lib/Demangle/MicrosoftDemangle.cpp
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module '/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/llvm/llvm/lib/Demangle/MicrosoftDemangle.cpp'.
4.	Running pass 'Complex Deinterleaving Pass' on function '@_ZL17guessCharByteSizePKhjm'
 #0 0x0000aaaaba931f20 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x8141f20)
 #1 0x0000aaaaba92fe6c llvm::sys::RunSignalHandlers() (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x813fe6c)
 #2 0x0000aaaaba896e80 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x0000ffff9e32c9d0 (linux-vdso.so.1+0x9d0)
 #4 0x0000ffff9de8f200 __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x0000ffff9de4a67c gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #6 0x0000ffff9de37130 abort ./stdlib/abort.c:81:7
 #7 0x0000ffff9de43fd0 __assert_fail_base ./assert/assert.c:89:7
 #8 0x0000ffff9de44040 __assert_perror_fail ./assert/assert-perr.c:31:1
 #9 0x0000aaaaba388dc8 llvm::Value::doRAUW(llvm::Value*, llvm::Value::ReplaceMetadataUses) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x7b98dc8)
#10 0x0000aaaab9c994b4 (anonymous namespace)::ComplexDeinterleavingGraph::replaceNode(llvm::IRBuilderBase&, (anonymous namespace)::ComplexDeinterleavingCompositeNode*) ComplexDeinterleavingPass.cpp:0:0
#11 0x0000aaaab9c8cefc (anonymous namespace)::ComplexDeinterleaving::evaluateBasicBlock(llvm::BasicBlock*) ComplexDeinterleavingPass.cpp:0:0
#12 0x0000aaaab9c89910 (anonymous namespace)::ComplexDeinterleaving::runOnFunction(llvm::Function&) ComplexDeinterleavingPass.cpp:0:0
#13 0x0000aaaab9c89ddc (anonymous namespace)::ComplexDeinterleavingLegacyPass::runOnFunction(llvm::Function&) ComplexDeinterleavingPass.cpp:0:0
#14 0x0000aaaaba31bfd0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x7b2bfd0)
#15 0x0000aaaaba323e7c llvm::FPPassManager::runOnModule(llvm::Module&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x7b33e7c)
#16 0x0000aaaaba31c950 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x7b2c950)
#17 0x0000aaaabb10eb3c clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x891eb3c)
#18 0x0000aaaabb13463c clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x894463c)
#19 0x0000aaaabccf960c clang::ParseAST(clang::Sema&, bool, bool) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0xa50960c)
#20 0x0000aaaabb52f25c clang::FrontendAction::Execute() (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x8d3f25c)
#21 0x0000aaaabb4b64b4 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x8cc64b4)
#22 0x0000aaaabb61c204 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x8e2c204)
#23 0x0000aaaab8207aac cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x5a17aac)
#24 0x0000aaaab82045f4 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#25 0x0000aaaabb361460 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::$_0>(long) Job.cpp:0:0
#26 0x0000aaaaba896b84 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x80a6b84)
#27 0x0000aaaabb3608b4 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x8b708b4)
#28 0x0000aaaabb326af0 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x8b36af0)
#29 0x0000aaaabb326ce4 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x8b36ce4)
#30 0x0000aaaabb33dd08 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x8b4dd08)
#31 0x0000aaaab82039e8 clang_main(int, char**, llvm::ToolContext const&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x5a139e8)
#32 0x0000aaaab8211e28 main (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x5a21e28)
#33 0x0000ffff9de373fc __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#34 0x0000ffff9de374cc call_init ./csu/../csu/libc-start.c:128:20
#35 0x0000ffff9de374cc __libc_start_main ./csu/../csu/libc-start.c:379:5
#36 0x0000aaaab82021f0 _start (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin/clang+++0x5a121f0)
clang++: error: clang frontend command failed with exit code 134 (use -v to see invocation)
clang version 20.0.0git (https://github.com/llvm/llvm-project.git e7303fe80a0bea124422219356c1c9e845110a77)
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/tcwg-buildbot/worker/clang-aarch64-sve-vla-2stage/stage1.install/bin
```

**Comment by antmox** [comment]
Also causing same failures here : 
https://lab.llvm.org/buildbot/#/builders/198/builds/334
https://lab.llvm.org/buildbot/#/builders/143/builds/4276
https://lab.llvm.org/buildbot/#/builders/17/builds/4725

**Comment by antmox** [comment]
And
https://lab.llvm.org/buildbot/#/builders/41/builds/4172
https://lab.llvm.org/buildbot/#/builders/4/builds/4281
https://lab.llvm.org/buildbot/#/builders/199/builds/263

**Comment by fhahn** [comment]
Reverted for now in 76714be5fd4ace66dd9e19ce706c2e2149dd5716 as this has been breaking most stage2 builds on AArch64 for 4+ hours

**Comment by NickGuy-Arm** [comment]
> Reverted for now in [76714be](https://github.com/llvm/llvm-project/commit/76714be5fd4ace66dd9e19ce706c2e2149dd5716) as this has been breaking most stage2 builds on AArch64 for 4+ hours

Thanks, I'm now testing a fix with a local stage2 build. Not sure why I didn't receive the buildbot emails though, so I didn't realise it was failing things at first.

**Comment by NickGuy-Arm** [comment]
Opened a new PR at #120441 to reland this patch with a fix for the build failures.

File: llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp [hunk 10]
```diff
@@ -1556,13 +1726,41 @@ void ComplexDeinterleavingGraph::identifyReductionNodes() {
         break;
       }
     }
+
+    auto *Real = OperationInstruction[i];
+    // We want to check that we have 2 operands, but the function attributes
+    // being counted as operands bloats this value.
+    if (Real->getNumOperands() < 2)
```
Comment (fhahn) [inline llvm/lib/CodeGen/ComplexDeinterleavingPass.cpp:1733]: Looks like we are missing a check to ensure we only have integer types here. With FP types, we currently crash due to creating `llvm.reduce.add` reductions for floating point types. It also doesn't check for the rigth FP flags AFAICT.

Put up https://github.com/llvm/llvm-project/pull/139469 to bail out

**Pull request merged**